src/HydraAuction/Platform/Interface.hs

```diff
@@ -0,0 +1,387 @@
+{-
+
+Entities are relation-like data, to be stored and queried in DB-like storage.
+
+Entity class describes their queriable behaviour.
+
+EntityKind is singletone-like functionalization for Entities,
+used to uniform queries and storage access.
+
+SomeX datatypes use existencial types to wrap EntityKind usage
+and make possible to use single datatype to access different Entities,
+as is required for server API.
+
+GADT usage makes deriving not working which affect some design decisions.
+-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module HydraAuction.Platform.Interface (
+  Entity (..),
+  EntityFilter (..),
+  AnnouncedAuction (..),
+  HydraHead (..),
+  HydraHeadInfo (..),
+  FilterEq (..),
+  FilterOrd (..),
+  HeadDelegate (..),
+  BidderApproval (..),
+  EntityKind (..),
+  ServerOutput (..),
+  Some (..),
+  ClientCommand (..),
+  CommandResult (..),
+  ClientInput (..),
+  EntityQueryResponse (..),
+  EntityQuery (..),
+) where
+
+-- Prelude imports
+import Prelude
+
+-- Haskell imports
+
+import Data.Aeson (FromJSON (..), ToJSON (..))
+import Data.Aeson qualified as Aeson
+import Data.Aeson.Types qualified as Aeson
+import Data.Kind (Type)
+import Data.Proxy (Proxy (..))
+import Data.Text qualified as T
+import GHC.Generics (Generic)
+
+-- Plutus imports
+import PlutusLedgerApi.V2 (BuiltinByteString)
+
+-- Hydra imports
+
+import Hydra.Cardano.Api (Lovelace)
+import Hydra.Chain (HeadId (..))
+
+-- HydraAuction imports
+
+import HydraAuction.Addresses (StandingBidAddress, VoucherCS)
+import HydraAuction.Delegate.Interface (
+  InitializedState,
+  InitializedStateKind,
+  initializedStateKind,
+ )
+import HydraAuction.Types (AuctionTerms)
+import HydraAuctionUtils.Fixture (Actor)
+import HydraAuctionUtils.Types.Natural (Natural)
+
+-- Interfaces
+
+type DefaultConstraints x = (Eq x, Show x, ToJSON x, FromJSON x, Generic x)
+
+class
+  ( Ord (PrimaryKey entity)
+  , DefaultConstraints (EntityFilter entity)
+  , DefaultConstraints (PrimaryKey entity)
+  , DefaultConstraints entity
+  ) =>
+  Entity entity
+  where
+  type PrimaryKey entity
+  data EntityFilter entity
+
+  getPrimaryKey :: entity -> PrimaryKey entity
+  isMatchingEntityFilter :: entity -> EntityFilter entity -> Bool
+
+-- Entities
+
+data AnnouncedAuction = MkAnnouncedAuction
+  { terms :: AuctionTerms
+  , auctionId :: VoucherCS
+  , -- Required for Delegate Servers
+    auctionStandingBidAddress :: StandingBidAddress
+  }
+  deriving stock (Generic, Eq, Ord, Show)
+  deriving anyclass (ToJSON, FromJSON)
+
+data HydraHeadInfo = MkHydraHeadInfo
+  { headId :: HeadId
+  , delegatesNumber :: Natural
+  , auctionFeePerDelegate :: Lovelace
+  }
+  deriving stock (Generic, Show, Eq, Ord)
+  deriving anyclass (ToJSON, FromJSON)
+
+data HydraHead = MkHydraHead
+  { staticInfo :: HydraHeadInfo
+  , allDelegatesKnown :: Bool
+  , headDelegateState :: InitializedState
+  }
+  deriving stock (Generic, Show, Eq)
+  deriving anyclass (ToJSON, FromJSON)
+
+deriving stock instance Ord HeadId
```

Comment:
Why is this derived separately?

Reply:
This is orphan instance for Hydra. We need it to worm map.

Could separate that to utils module or something like that.

Reply:
Oh sorry, thought it was for `HydraHead` :/

Reply:
This is orphan instance for Hydra. We need it to worm map.

Could separate that to utils module or something like that.